File: C#/Solver.cs
using System;
using System.Collections.Generic;
class Solver
{
    public static int[][] Solve(int[][] board)
    {
        (int y, int x) = find_next(board, 0, 0);
        if (x == -1 || y == -1)
        {
            return new int[][] {};
        }
        return rec_solve(board, x, y, successors(board, x, y));
    }

    static private int[][] rec_solve(int[][] board, int x, int y, List<int> successors_list){
        foreach (int successor in successors_list){
            board[y][x] = successor;
            int new_x = x+1;
            int new_y = y;
            if (new_x == 9){
                new_x = 0;
                new_y++;
            }
            (new_y, new_x) = find_next(board, new_x, new_y);
            if (new_x == -1 || new_y == -1){
                return board;
            }
            List<int> next = successors(board, new_x, new_y);
            if (next.Count == 0){
                board[y][x] = 0;
                continue;
            }
            int [][] result = rec_solve(board, new_x, new_y, next);
            if (result.Length != 0){
                return result;
            }
            board[y][x] = 0;
        }
        return new int[][] {};
    }
    static private List<int> successors(int[][] board, int x, int y){
        if (board[y][x] != 0){
            (y,x) = find_next(board, x, y);
        }
        if (x == -1 || y == -1){
            return new List<int>();
        }
        List<int> x_row = new List<int>(board[y]);
        List<int> y_row = new List<int>();
        for (int i = 0; i < board.Length; i++){
            y_row.Add(board[i][x]);
        }
        List<int> block = new List<int>();
        for (int i = 0; i < 3; i++){
            int block_y = (y / 3) * 3 + i;
            for (int j = 0; j < 3; j++){
                int block_x = (x / 3) * 3 + j;
                block.Add(board[block_y][block_x]);
            }
        }
        List<int> output = new List<int>();
        for (int num = 1; num < 10; num++)
        {
            if (!x_row.Contains(num) && !y_row.Contains(num) && !block.Contains(num))
            {
                output.Add(num);
            }
        }
        return output;
    }

    static private (int y, int x) find_next(int[][] board, int x, int y){
        for (int i = y; i <board.Length; i++)
        {
            for (int j = x; j < board[i].Length; j++)
            {
                if (board[i][j] == 0)
                {
                    return (i, j);
                }
            }
            x = 0;
        }
        return (-1, -1);
    }
}

File: Kotlin/src/main/kotlin/Solver.kt
public fun Solve(Board: Array<Array<Int>>): Array<Array<Int>> {
    var x = 0
    var y = 0
    if (Board[y][x] != 0) {
        val pair = findNext(Board, x, y)
        x = pair.first
        y = pair.second
        if (x == -1 || y == -1){
            return emptyArray()
        }
    }
    val successorsList = successors(Board, x, y)
    return recSolve(Board, x, y, successorsList)
}

private fun recSolve(Board: Array<Array<Int>>, x : Int, y : Int, successorsList: List<Int>): Array<Array<Int>>{

    for (successor in successorsList){
        Board[y][x] = successor
        var newX = x + 1
        var newY = y
        if (newX == 9){
            newY++
            newX = 0
        }
        val new = findNext(Board, newX, newY)
        newX = new.first
        newY = new.second
        if (newX == -1 || newY == -1){
            return Board
        }
        val next = successors(Board, newX, newY)
        if (next.isEmpty()){
            Board[y][x] = 0
            continue
        }
        val result = recSolve(Board, newX, newY, next)
        if (result.isNotEmpty()){
            return result
        }
        Board[y][x] = 0
    }
    return emptyArray()
}

private fun successors(Board: Array<Array<Int>>, x: Int, y : Int): List<Int> {

    val xRow = Board[y]
    val yRow = arrayOfNulls<Int>(9)
    for (i in Board.indices){
        yRow[i] = Board[i][x]
    }
    var block = ArrayList<Int>()
    for (i in 0..2){
        val blockY = (y / 3) * 3 + i
        for (j in 0..2){
            val blockX = (x / 3) * 3 + j
            block.add(Board[blockY][blockX])
        }
    }
    var output = ArrayList<Int>()
    for (num in 1..9){
        if (!xRow.contains(num) && !yRow.contains(num) && !block.contains(num)){
            output.add(num)
        }
    }
    return output
}

private fun findNext(Board: Array<Array<Int>>, x: Int, y: Int): Pair<Int, Int>{
    var tempX = x
    for (i in y until Board.size){
        for (j in tempX until Board[i].size){
            if (Board[i][j] == 0){
                return Pair(j, i)
            }
        }
        tempX = 0
    }
    return Pair (-1, -1)
}